usr/src/uts/common/fs/zfs/vdev_removal.c

```diff
@@ -1089,39 +1209,75 @@ spa_vdev_copy_impl(vdev_t *vd, spa_vdev_removal_t *svr, vdev_copy_arg_t *vca,
 
 	mutex_enter(&svr->svr_lock);
 
-	range_seg_t *rs = avl_first(&svr->svr_allocd_segs->rt_root);
-	if (rs == NULL) {
+	if (range_tree_is_empty(svr->svr_allocd_segs)) {
 		mutex_exit(&svr->svr_lock);
 		return;
 	}
-	uint64_t offset = rs->rs_start;
-	uint64_t length = MIN(rs->rs_end - rs->rs_start, *max_alloc);
 
-	range_tree_remove(svr->svr_allocd_segs, offset, length);
+	/*
+	 * Determine how big of a chunk to copy.  We can allocate up
+	 * to max_alloc bytes, and we can span up to vdev_removal_max_span
+	 * bytes of unallocated space at a time.  "segs" will track the
+	 * allocated segments that we are copying.  We may also be copying
+	 * free segments (of up to vdev_removal_max_span bytes).
+	 */
+	range_tree_t *segs = range_tree_create(NULL, NULL);
+	for (;;) {
+		range_seg_t *rs = avl_first(&svr->svr_allocd_segs->rt_root);
```

Comment:
Testing my WIP port to ZoL, I'm seeing "rs == NULL".  I've put in an ASSERT below the `range_tree_is_empty()` above to make sure `avl_first()` is not null when empty but somehow, as part of the loop, `avl_first()` is able to return NULL.  I'm trying to audit all the uses of `svr_allocd_segs` to make sure they're all being done under the `svr_lock`.

Mentioning @behlendorf so he knows I'm working on porting this.